File: fsharp/dnd-character/DndCharacter.fs
module DndCharacter

open System

let modifier x = 
    let calc a = (a - 10.0) / 2.0
    x |> float |> calc |> floor |> int

let ability() = 
    let rnd = Random()
    List.init 4 (fun _ -> rnd.Next(1, 6)) 
    |> List.sortDescending 
    |> List.take 3 
    |> List.sum

type Character = {
    Strength: int;
    Dexterity: int;
    Constitution: int;
    Intelligence: int;
    Wisdom: int;
    Charisma: int;
    Hitpoints: int;
}

let createCharacter() =
    let constitution = ability()
    let hp = 10 + modifier constitution
    { 
        Strength=ability(); 
        Dexterity=ability(); 
        Constitution=constitution; 
        Intelligence=ability(); 
        Wisdom=ability(); 
        Charisma=ability(); 
        Hitpoints=hp
    }

File: python/dnd-character/dnd_character.py
import math
import random

def modifier(n):
    return math.floor((n - 10) / 2)

class Character:
    def __init__(self):
        self.strength = self.ability()
        self.dexterity = self.ability()
        self.constitution = self.ability()
        self.intelligence = self.ability()
        self.wisdom = self.ability()
        self.charisma = self.ability()
        self.hitpoints = 10 + modifier(self.constitution)

    def ability(self):
        rolls = random.sample(range(1, 7), 4)
        return sum(sorted(rolls, reverse=True)[0:3])
